cpp/include/kvikio/remote_handle.hpp

```diff
@@ -145,12 +146,178 @@ class HttpEndpoint : public RemoteEndpoint {
+  std::string _aws_sigv4;
+  std::string _aws_userpwd;
+
+  static std::string parse_aws_argument(std::optional<std::string> aws_arg,
+                                        std::string const& env_var,
+                                        std::string const& err_msg,
+                                        bool allow_empty = false)
+  {
+    if (aws_arg.has_value()) { return std::move(*aws_arg); }
+
+    char const* env = std::getenv(env_var.c_str());
+    if (env == nullptr) {
+      if (allow_empty) { return std::string(); }
+      throw std::invalid_argument(err_msg);
+    }
+    return std::string(env);
+  }
+
+  static std::string url_from_bucket_and_object(std::string const& bucket_name,
+                                                std::string const& object_name,
+                                                std::optional<std::string> const& aws_region,
+                                                std::optional<std::string> aws_endpoint_url)
+  {
+    std::string endpoint_url =
+      parse_aws_argument(std::move(aws_endpoint_url), "AWS_ENDPOINT_URL", "", true);
+    std::stringstream ss;
+    if (endpoint_url.empty()) {
+      std::string region =
+        parse_aws_argument(std::move(aws_region),
+                           "AWS_DEFAULT_REGION",
+                           "S3: must provide `aws_region` if AWS_DEFAULT_REGION isn't set.");
+      // We default to the official AWS url scheme.
+      ss << "https://" << bucket_name << ".s3." << region << ".amazonaws.com/" << object_name;
+    } else {
+      ss << endpoint_url << "/" << bucket_name << "/" << object_name;
+    }
+    return ss.str();
+  }
+
+ public:
+  /**
+   * @brief Given an url like "s3://<bucket>/<object>", return the name of the bucket and object.
+   *
+   * @throws std::invalid_argument if url is ill-formed or is missing the bucket or object name.
+   *
+   * @param s3_url S3 url.
+   * @return Pair of strings: [bucket-name, object-name].
+   */
+  [[nodiscard]] static std::pair<std::string, std::string> parse_s3_url(std::string const& s3_url)
+  {
+    if (s3_url.empty()) { throw std::invalid_argument("The S3 url cannot be an empty string."); }
+    if (s3_url.size() < 5 || s3_url.substr(0, 5) != "s3://") {
+      throw std::invalid_argument("The S3 url must start with the S3 scheme (\"s3://\").");
+    }
+    std::string p = s3_url.substr(5);
+    if (p.empty()) { throw std::invalid_argument("The S3 url cannot be an empty string."); }
+    size_t pos              = p.find_first_of('/');
+    std::string bucket_name = p.substr(0, pos);
+    if (bucket_name.empty()) {
+      throw std::invalid_argument("The S3 url does not contain a bucket name.");
+    }
+    std::string object_name = (pos == std::string::npos) ? "" : p.substr(pos + 1);
+    if (object_name.empty()) {
+      throw std::invalid_argument("The S3 url does not contain an object name.");
+    }
+    return std::make_pair(std::move(bucket_name), std::move(object_name));
```

Comment:
This feels like a lot of work to avoid a regex :joy: 
```suggestion
    // Regular expression to match s3://<bucket>/<object>
    std::regex pattern{R"(s3://([^/]+)/(.+))"};
    std::smatch matches;
    if (std::regex_match(s3_url, matches, pattern)) {
      return {matches[1].str(), matches[2].str()};
    }
    throw std::invalid_argument("Input string does not match the expected S3 URL format.");
```
You'll have to `#include <regex>` as well.

Reply:
Heh, thanks :)